lighthouse-core/audits/seo/canonical.js

```diff
@@ -0,0 +1,166 @@
+      requiredArtifacts: ['Canonical', 'Hreflang'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+
+    return artifacts.requestMainResource(devtoolsLogs)
+      .then(mainResource => {
+        const baseURL = new URL(mainResource.url);
+        let canonicals = [];
+        let hreflangs = [];
+
+        mainResource.responseHeaders
+          .filter(h => h.name.toLowerCase() === LINK_HEADER)
+          .forEach(h => {
+            canonicals = canonicals.concat(getCanonicalLinksFromHeader(h.value));
+            hreflangs = hreflangs.concat(getHreflangsFromHeader(h.value));
+          });
+
+        canonicals = canonicals.concat(artifacts.Canonical);
+
+        artifacts.Hreflang.forEach(({href}) => hreflangs.push(href));
+
+        hreflangs = hreflangs
+          .filter(href => isValidRelativeOrAbsoluteURL(href))
+          .map(href => (new URL(href, baseURL)).href); // normalize URLs
+
+        if (canonicals.length === 0) {
+          return {
+            rawValue: true,
+          };
+        }
+
+        if (canonicals.length > 1) {
+          return {
+            rawValue: false,
+            displayValue: `multiple URLs (${canonicals.join(', ')})`,
+          };
+        }
+
+        const canonical = canonicals[0];
+
+        if (!isValidRelativeOrAbsoluteURL(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `invalid URL (${canonical})`,
+          };
+        }
+
+        if (!URL.isValid(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `relative URL (${canonical})`,
+          };
+        }
+
+        const canonicalURL = new URL(canonical);
+
+        // cross-language or cross-country canonicals are a common issue
+        if (hreflangs.includes(baseURL.href) && hreflangs.includes(canonicalURL.href) &&
+          baseURL.href !== canonicalURL.href) {
+          return {
+            rawValue: false,
+            displayValue: `points to another hreflang location (${baseURL.href})`,
+          };
+        }
+
+        // bing and yahoo don't allow canonical URLs pointing to different domains, it's also
+        // a common mistake to publish a page with canonical pointing to e.g. a test domain or localhost
+        if (getPrimaryDomain(canonicalURL) !== getPrimaryDomain(baseURL)) {
+          return {
+            rawValue: false,
+            displayValue: `points to a different domain (${canonicalURL})`,
+          };
+        }
+
+        // another common mistake is to have canonical pointing from all pages of the website to its root
+        if (canonicalURL.origin === baseURL.origin &&
+          canonicalURL.pathname === '/' && baseURL.pathname !== '/') {
+          return {
+            rawValue: false,
+            displayValue: 'points to a root of the same origin',
```

Comment:
@rviscomi I made these `displayValue`s very short and while this works for `invalid URL` all others require more explanation. Not sure if `displayValue` is the right place for this explanation though (limited formatting, not much space) - maybe separate documentation page that we discussed would help here?

Reply:
Can the `failureDescription` be overwritten to have failure-specific explanations?

Reply:
From what I tested neither `failureDescription` nor `helpText` can be overwritten ☹️ We only have `displayValue` and `details`. Unfortunately, `details` do not support any formatting when it comes to text:

<img width="657" alt="screen shot 2018-01-03 at 02 52 37" src="https://user-images.githubusercontent.com/985504/34506932-62c85df8-f031-11e7-9070-4f4ff232e5c6.png">

Reply:
Is the immutability of `failureDescription` a bug or a feature? cc @paulirish